### PR TITLE
CATROID-1132 Refactor CategoryBricksFactory

### PR DESCRIPTION
**Pull request by Walcho1125**
*Please enter a short description of your pull request and add a reference to the Jira ticket.*
https://jira.catrob.at/browse/CATROID-1132

Refactor CategoryBricksFactory.java to Kotlin
### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [ ] Verify that your changes do not have any conflicts with the base branch
- [ ] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

**Review by dandrawes — APPROVED** [review]
Well done!

File: catroid/src/main/java/org/catrobat/catroid/ui/fragment/CategoryBricksFactory.kt [hunk 1]
```diff
@@ -0,0 +1,777 @@
...
+        chromecastBrickList.add(WhenGamepadButtonBrick(WhenGamepadButtonScript(context.getString(R.string.cast_gamepad_A))))
+        return chromecastBrickList
+    }
+
+    protected open fun setupMotionCategoryList(
+        context: Context?,
+        isBackgroundSprite: Boolean
+    ): List<Brick> {
+        val motionBrickList: MutableList<Brick> = ArrayList()
+        motionBrickList.add(PlaceAtBrick(BrickValues.X_POSITION, BrickValues.Y_POSITION))
```
Comment (gPathpp) [inline catroid/src/main/java/org/catrobat/catroid/ui/fragment/CategoryBricksFactory.kt:374]: Try to use addAll, please.

File: catroid/src/main/java/org/catrobat/catroid/ui/fragment/CategoryBricksFactory.kt [hunk 2]
```diff
@@ -0,0 +1,777 @@
...
+        dataBrickList.add(AddItemToUserListBrick(BrickValues.ADD_ITEM_TO_USERLIST))
+        dataBrickList.add(DeleteItemOfUserListBrick(BrickValues.DELETE_ITEM_OF_USERLIST))
+        dataBrickList.add(ClearUserListBrick())
+        dataBrickList.add(InsertItemIntoUserListBrick(BrickValues.INSERT_ITEM_INTO_USERLIST_VALUE, BrickValues.INSERT_ITEM_INTO_USERLIST_INDEX))
+        dataBrickList.add(ReplaceItemInUserListBrick(BrickValues.REPLACE_ITEM_IN_USERLIST_VALUE, BrickValues.REPLACE_ITEM_IN_USERLIST_INDEX))
+        dataBrickList.add(WriteListOnDeviceBrick())
+        dataBrickList.add(ReadListFromDeviceBrick())
+        dataBrickList.add(StoreCSVIntoUserListBrick(BrickValues.STORE_CSV_INTO_USERLIST_COLUMN, context.getString(R.string.brick_store_csv_into_userlist_data)))
+        dataBrickList.add(WebRequestBrick(context.getString(R.string.brick_web_request_default_value)))
+        when {
```
Comment (gPathpp) [inline catroid/src/main/java/org/catrobat/catroid/ui/fragment/CategoryBricksFactory.kt:533]: Nice

File: catroid/src/main/java/org/catrobat/catroid/ui/fragment/CategoryBricksFactory.kt [hunk 3]
```diff
@@ -0,0 +1,777 @@
...
+        val deviceBrickList: MutableList<Brick> = ArrayList()
+        deviceBrickList.add(ResetTimerBrick())
+        deviceBrickList.add(WhenBrick())
+        deviceBrickList.add(WhenTouchDownBrick())
+        if (SettingsFragment.isNfcSharedPreferenceEnabled(context)) {
+            deviceBrickList.add(WhenNfcBrick())
+            deviceBrickList.add(SetNfcTagBrick(context.getString(R.string.brick_set_nfc_tag_default_value)))
+        }
+        deviceBrickList.add(WebRequestBrick(context.getString(R.string.brick_web_request_default_value)))
+        when {
```
Comment (gPathpp) [inline catroid/src/main/java/org/catrobat/catroid/ui/fragment/CategoryBricksFactory.kt:565]: Please remove duplicated code. 
Add a function or smth similar. :)

File: catroid/src/main/java/org/catrobat/catroid/ui/fragment/CategoryBricksFactory.kt [hunk 4]
```diff
@@ -0,0 +1,777 @@
...
+        assertionsBrickList.add(TapForBrick(BrickValues.TOUCH_X_START, BrickValues.TOUCH_Y_START, BrickValues.TOUCH_DURATION))
+        assertionsBrickList.add(TouchAndSlideBrick(BrickValues.TOUCH_X_START, BrickValues.TOUCH_Y_START, BrickValues.TOUCH_X_GOAL, BrickValues.TOUCH_Y_GOAL, BrickValues.TOUCH_DURATION))
+        assertionsBrickList.add(FinishStageBrick())
+        assertionsBrickList.add(StoreCSVIntoUserListBrick(BrickValues.STORE_CSV_INTO_USERLIST_COLUMN, context.getString(R.string.brick_store_csv_into_userlist_data)))
+        assertionsBrickList.add(WebRequestBrick(context.getString(R.string.brick_web_request_default_value)))
+        return assertionsBrickList
+    }
+
+    fun searchList(searchBrick: Brick, list: List<Brick>): Boolean {
+        list.forEach { brick ->
```
Comment (gPathpp) [inline catroid/src/main/java/org/catrobat/catroid/ui/fragment/CategoryBricksFactory.kt:722]: https://stackoverflow.com/questions/57707276/kotlin-check-if-list-has-an-object-of-a-specific-type

File: catroid/src/main/java/org/catrobat/catroid/ui/fragment/CategoryBricksFactory.kt [hunk 5]
```diff
@@ -0,0 +1,777 @@
...
+    }
+
+    fun getBrickCategory(brick: Brick, isBackgroundSprite: Boolean, context: Context): String? {
+        val res = context.resources
+        val config = res.configuration
+        val savedLocale = config.locale
+        config.locale = Locale.ENGLISH
+        res.updateConfiguration(config, null)
+        var category = "No match"
+        when {
```
Comment (gPathpp) [inline catroid/src/main/java/org/catrobat/catroid/ui/fragment/CategoryBricksFactory.kt:737]:
```suggestion
        category = when {
```

**Review by gPathpp — CHANGES_REQUESTED** [review]
LGTM, just some small code improvements.

File: catroid/src/main/java/org/catrobat/catroid/ui/fragment/CategoryBricksFactory.kt (hunk 3 above)
Comment (Walcho1125) [reply catroid/src/main/java/org/catrobat/catroid/ui/fragment/CategoryBricksFactory.kt:561]: will be done in future refactoring

File: catroid/src/main/java/org/catrobat/catroid/ui/fragment/CategoryBricksFactory.kt [hunk 6]
```diff
@@ -0,0 +1,767 @@
...
+        looksBrickList.add(ChangeBrightnessByNBrick(BrickValues.CHANGE_BRIGHTNESS_BY))
+        looksBrickList.add(SetColorBrick(BrickValues.SET_COLOR_TO))
+        looksBrickList.add(ChangeColorByNBrick(BrickValues.CHANGE_COLOR_BY))
+        looksBrickList.add(ClearGraphicEffectBrick())
+        looksBrickList.add(WhenBackgroundChangesBrick())
+        looksBrickList.add(SetBackgroundBrick())
+        looksBrickList.add(SetBackgroundByIndexBrick(BrickValues.SET_LOOK_BY_INDEX))
+        looksBrickList.add(SetBackgroundAndWaitBrick())
+        looksBrickList.add(SetBackgroundByIndexAndWaitBrick(BrickValues.SET_LOOK_BY_INDEX))
+        if (!ProjectManager.getInstance().currentProject.isCastProject) {
```
Comment (ndrnour) [inline catroid/src/main/java/org/catrobat/catroid/ui/fragment/CategoryBricksFactory.kt:472]: please use injected ProjectManager
```suggestion
        if (!projectManager.currentProject.isCastProject) {
```

File: catroid/src/main/java/org/catrobat/catroid/ui/fragment/CategoryBricksFactory.kt [hunk 7]
```diff
@@ -0,0 +1,767 @@
...
+            looksBrickList.add(CameraBrick())
+            looksBrickList.add(ChooseCameraBrick())
+            looksBrickList.add(FlashBrick())
+        }
+        when {
+            !isBackgroundSprite -> looksBrickList.add(LookRequestBrick(BrickValues.LOOK_REQUEST))
+            projectManager.currentProject.xmlHeader.islandscapeMode() -> looksBrickList.add(BackgroundRequestBrick(BrickValues.BACKGROUND_REQUEST_LANDSCAPE))
```
Comment (ndrnour) [inline catroid/src/main/java/org/catrobat/catroid/ui/fragment/CategoryBricksFactory.kt:479]: same

**Review by ndrnour — CHANGES_REQUESTED** [review]
please use the injected ProjectManager, otherwise, this would cause some test cases to fail

File: catroid/src/main/java/org/catrobat/catroid/ui/fragment/CategoryBricksFactory.kt [hunk 8]
```diff
@@ -0,0 +1,767 @@
...
+        assertionsBrickList.add(TouchAndSlideBrick(BrickValues.TOUCH_X_START, BrickValues.TOUCH_Y_START, BrickValues.TOUCH_X_GOAL, BrickValues.TOUCH_Y_GOAL, BrickValues.TOUCH_DURATION))
+        assertionsBrickList.add(FinishStageBrick())
+        assertionsBrickList.add(StoreCSVIntoUserListBrick(BrickValues.STORE_CSV_INTO_USERLIST_COLUMN, context.getString(R.string.brick_store_csv_into_userlist_data)))
+        assertionsBrickList.add(WebRequestBrick(context.getString(R.string.brick_web_request_default_value)))
+        return assertionsBrickList
+    }
+
+    fun searchList(searchBrick: Brick, list: List<Brick>): Boolean = list.any { it == searchBrick.javaClass }
+
+    fun getBrickCategory(brick: Brick, isBackgroundSprite: Boolean, context: Context): String? {
```
Comment (ndrnour) [inline catroid/src/main/java/org/catrobat/catroid/ui/fragment/CategoryBricksFactory.kt:719]: why nullable string as return type?
```suggestion
    fun getBrickCategory(brick: Brick, isBackgroundSprite: Boolean, context: Context): String {
```

File: catroid/src/main/java/org/catrobat/catroid/ui/fragment/CategoryBricksFactory.kt [hunk 9]
```diff
@@ -0,0 +1,767 @@
...
+        soundBrickList.add(PlayNoteForBeatsBrick(BrickValues.DEFAULT_NOTE, BrickValues.PAUSED_BEATS_INT))
+        soundBrickList.add(PlayDrumForBeatsBrick(BrickValues.PAUSED_BEATS_INT))
+        soundBrickList.add(SetTempoBrick(BrickValues.DEFAULT_TEMPO))
+        soundBrickList.add(ChangeTempoByNBrick(BrickValues.CHANGE_TEMPO))
+        soundBrickList.add(PauseForBeatsBrick(BrickValues.PAUSED_BEATS_FLOAT))
+        return soundBrickList
+    }
+
+    protected open fun setupLooksCategoryList(context: Context, isBackgroundSprite: Boolean): List<Brick> {
+        val looksBrickList: MutableList<Brick> = ArrayList()
```
Comment (ndrnour) [inline catroid/src/main/java/org/catrobat/catroid/ui/fragment/CategoryBricksFactory.kt:440]: can be simplified as 
```suggestion
        val looksBrickList = mutableListOf<Brick>()
```

**Comment by ndrnour** [comment]
APKs build fails please fix it too. see [log](https://jenkins.catrob.at/job/Catroid/job/PR-4215/26/execution/node/39/log/)

**Review by ndrnour — CHANGES_REQUESTED** [review]
in order to fix the new failing test, you need to use injected projectManager in `ActionFactory.java`

**Review by ndrnour — APPROVED** [review]

**Review by gPathpp — APPROVED** [review]

**Comment by ndrnour** [comment]
thanks!

**Pull request merged**